src/main/java/org/broadinstitute/dsm/cf/TestBostonKitTrackerDispatcher.java

```diff
@@ -0,0 +1,134 @@
+                "k.tracking_to_id,\n" +
+                "k.tracking_return_id,\n" +
+                "json_extract(k.test_result, '$[0].result') test_result,\n" +
+                "ifnull(STR_TO_DATE(replace(json_extract(k.test_result, '$[0].timeCompleted'),'\"',''), '%Y-%m-%dT%H:%i:%sZ'),\n" +
+                "    STR_TO_DATE(replace(json_extract(k.test_result, '$[0].timeCompleted'),'\"',''), '%Y-%m-%dT%H:%i:%s.%fZ')\n" +
+                ")as test_completion_time,\n" +
+                "u.hruid,\n" +
+                "dkr.upload_reason\n" +
+                "from\n" +
+                STUDY_MANAGER_SCHEMA + "ddp_kit k,\n" +
+                STUDY_MANAGER_SCHEMA + "ddp_kit_request dkr,\n" +
+                STUDY_SERVER_SCHEMA + "user u,\n" +
+                STUDY_MANAGER_SCHEMA + "ddp_instance study,\n" +
+                STUDY_MANAGER_SCHEMA + "kit_type kt\n" +
+                "where\n" +
+                "study.instance_name = 'testboston'\n" +
+                "and\n" +
+                "u.guid = dkr.ddp_participant_id\n" +
+                "and\n" +
+                "study.ddp_instance_id = dkr.ddp_instance_id\n" +
+                "and\n" +
+                "dkr.dsm_kit_request_id = k.dsm_kit_request_id\n" +
+                "and\n" +
+                "kt.kit_type_id = dkr.kit_type_id\n" +
+                "and\n" +
+                "kt.kit_type_name = 'AN'\n" +
+                "and k.kit_label like 'TBOS-%'\n" +
+                "order by test_completion_time";
+
+
+        logger.info("Starting up in project " + projectId + " and secret " + secretId);
+        Config cfg = null;
+
+        try (SecretManagerServiceClient secretManagerServiceClient = SecretManagerServiceClient.create()) {
+            var latestSecretVersion = SecretVersionName.of(projectId, secretId, "latest");
+            SecretPayload secret = secretManagerServiceClient.accessSecretVersion(latestSecretVersion).getPayload();
+            String secretData = secret.getData().toStringUtf8();
+            cfg = ConfigFactory.parseString(secretData);
+        }
+        String dbUrl = cfg.getString("pepperDbUrl");
+
+        PoolingDataSource<PoolableConnection> dataSource = createDataSource(1, dbUrl);
+
+        // static vars are dangerous in CF https://cloud.google.com/functions/docs/bestpractices/tips#functions-tips-scopes-java
+        try (Connection conn = dataSource.getConnection()) {
+            try (PreparedStatement stmt = conn.prepareStatement(KIT_QUERY)) {
+                stmt.setFetchSize(1000);
+                try (ResultSet rs = stmt.executeQuery()) {
+                    while (rs.next()) {
+                        String userGuid = rs.getString("guid");
+                        logger.info(userGuid);
+
+                        // spawn threads or other cloud functions in fixed size blocks? Write to a separate queue for worker threads
+                        // to grab history?
```

Comment:
With CF limits, churning through the whole result set here will result in termination before the list is complete.  I'm leaning towards capturing these guids as elements of a generic worklist in json as a sort of IPC for this and other producer/consumer patterns.

A TBD class like `TestBostonKitTrackingConsumer` or some such could then run on a GCP schedule (let's say every 30m) and run through the queue, keeping track of its own execution time, shutting down gracefully before it is terminated.  In this way we would have sustained progress through the list without running afoul of CF limits, and without overwhelming UPS with many concurrent calls.